improved QSD synthesis algorithm

ShellyGarion opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
This PR improves the QSD synthesis algorithm based on the Block ZXZ-Decomposition method described in the paper https://arxiv.org/abs/2403.13692v2.
This PR is aimed to be part of porting the QS decomposition code into rust #12213.

### Details and comments
We should consider the API - the parameters "opt_a1" and "opt_a2" should be removed.

Here are a few benchmarks. I took a random unitary matrix on `num_qubits=(3,4,5,6)` and `num_controls=(0,1,2,3)` and tried the following synthesis algorithms (transpiling with `optimization_level=0`):

1. `Isometry`  https://docs.quantum.ibm.com/api/qiskit/qiskit.circuit.library.Isometry
2. QSD with `opt_a1=False` and `opt_a2=False`
3. QSD with `opt_a1=True` and `opt_a2=True`
4. QSD with `opt_a1=True` and `opt_a2=False` (so we can do the multiplexed reduction for controlled gates)
5. QSD with ZXZ and `opt_a1=False` and `opt_a2=False`
6. QSD with ZXZ and `opt_a1=True` and `opt_a2=True`
7. QSD with ZXZ and `opt_a1=True` and `opt_a2=False`

Here are the CX counts results (the best result is in **bold**):

| num_qubits | num_controls | Isometry | QSD: `opt_a1=F`, `opt_a2=F` | QSD: `opt_a1=T`, `opt_a2=T` | QSD: `opt_a1=T`, `opt_a2=F` | ZXZ: `opt_a1=F`, `opt_a2=F` | ZXZ: `opt_a1=T`, `opt_a2=T` | ZXZ: `opt_a1=T`, `opt_a2=F` |
|--------------|----------------|-----------|-----------------------------------------|----------------------------------------|---------------------------------------------------|-----------------------------------------|----------------------------------------|---------------------------------------------------|
| 3 | 0 | 41 | 24 | 20 | 23 | 24 | **19** | 22 | 
| 3 | 1 | 144 | 56 | 99 | 54 | 56 | 95 | **52** |
| 3 | 2 | 922 | 252 | 444 | 244 | 254 | 423 | **235** |
| 3 | 3 | 1741 | 1088 | 1868 | 1046 | 1088 | 1783 | **1004** |
| 4 | 0 | 218 | 120 | 100 | 115 | 120 | **95** | 110 |  
| 4 | 1 | 928 | 256 | 444 | 246 | 256 | 423 | **236** |
| 4 | 2 | 4279 | 1088 | 1868 | 1046 | 1088 | 1783 | **1004** |
| 4 | 3 | 18348 | 4480 | 7660 | 4310 | 4480 | 7319 | **4140** | 
| 5 | 0 | 1025 | 528 | 444 | 507 | 528 | **423** | 486 |
| 5 | 1 | 2811 | 1088 | 1868 | 1046 | 1088 | 1783 | **1004** |
| 5 | 2 | 18131 | 4480 | 7660 | 4310 | 4478 | 7319 | **4138** | 
| 5 | 3 | 26022 | 18176 | 31020 | 17494 | 18176 | 29655 | **16812** |  
| 6 | 0 | 4474 | 2208 | 1868 | 2123 | 2208 | **1783** | 2038 |
| 6 | 1 | 11658 | 4480 | 7660 | 4310 | 4480 | 7319 | **4140** | 
| 6 | 2 | 72595 | 18176 | 31020 | 17494 | 18176 | 29655 | **16812** |

ShellyGarion committed:
add ZXZ decomposition

ShellyGarion committed:
merge wmatA B vmatC

ShellyGarion committed:
fix a misprint

ShellyGarion committed:
add ucrz trick

ShellyGarion committed:
formatting and docs

ShellyGarion committed:
fix failing tests

ShellyGarion committed:
add opt_a1 option

ShellyGarion committed:
update cx count upper bound

ShellyGarion committed:
remove UCRZGate class

ShellyGarion committed:
add closest unitary func to avoid M2 diagonalization errors

ShellyGarion committed:
update comment

ShellyGarion committed:
updtae documentatio and comments

ShellyGarion committed:
update tests

ShellyGarion committed:
add release notes

ShellyGarion committed:
fix misprints in docs

ShellyGarion committed:
add formula for final number of CX gates

gadial commented on qiskit/synthesis/unitary/qsd.py:
```diff
@@ -120,52 +139,120 @@ def decomp_2q(mat):
                 decomposer_2q = TwoQubitBasisDecomposer(CXGate())
+        mat = _closest_unitary(mat)
         circ = decomposer_2q(mat)
     else:
```
why use an elif structure when you can return the circ for dim==2 or dim==4 right after computing it?

gadial commented on qiskit/synthesis/unitary/qsd.py:
```diff
@@ -120,52 +139,120 @@ def decomp_2q(mat):
+
+    VX, S, WXdg = scipy.linalg.svd(X)
+    Sigma = np.zeros((n, n))
+    for i in range(n):
```
You can do `np.diag(S)`.

gadial commented on qiskit/synthesis/unitary/qsd.py:
```diff
@@ -120,52 +139,120 @@ def decomp_2q(mat):
+    Sigma = np.zeros((n, n))
+    for i in range(n):
+        Sigma[i, i] = S[i]
+    VXdg = scipy.linalg.inv(VX)
```
Since `VX` is unitary, isn't it better to do `VXdg = VX.conj().T`?

gadial commented on qiskit/synthesis/unitary/qsd.py:
```diff
@@ -120,52 +139,120 @@ def decomp_2q(mat):
+    Sigma = np.zeros((n, n))
+    for i in range(n):
+        Sigma[i, i] = S[i]
+    VYdg = scipy.linalg.inv(VY)
```
Same remarks as for VX

gadial commented on qiskit/synthesis/unitary/qsd.py:
```diff
@@ -120,52 +139,120 @@ def decomp_2q(mat):
+    for i in range(n):
+        Sigma[i, i] = S[i]
+    VXdg = scipy.linalg.inv(VX)
+    SX = np.dot(VX, np.dot(Sigma, VXdg))
```
`SX = VX @ Sigma @ VXdg` might be more readable

gadial commented on qiskit/synthesis/unitary/qsd.py:
```diff
@@ -120,52 +139,120 @@ def decomp_2q(mat):
+        # zmat is needed in order to reduce two cz gates, and combine them into the B2 matrix
+        zmat = np.eye(2 ** (nqubits - 1))
+        for i in range(2 ** (nqubits - 2), 2 ** (nqubits - 1)):
+            zmat[i][i] = -1
```
I think this can be done as one-liner `zmat = np.diag([1]*(dim // 4) + [-1]*(dim // 4))`

gadial commented on qiskit/synthesis/unitary/qsd.py:
```diff
@@ -221,56 +316,63 @@ def _demultiplex(um0, um1, opt_a1=False, opt_a2=False, *, _depth=0, _ctrl_index=
+            circuit.cx(q_controls[q_contr_index], q_target)
+        elif _vw_type == "all":
+            q_contr_index = len(q_controls) - 1
+            circuit.cx(q_controls[q_contr_index], q_target)
```
why repeat this line twice in each branch instead of putting in once after the branch?

ShellyGarion replied:
this is actually needed since it's not an if-else statement, but an if-elif 
(there are cases when adding cx is not done)

gadial replied:
Indeed! In `_dec_ucrot` we don't have `_vw_type`, missed that one.

gadial reviewed (dismissed):
While I can't say I 100% understand the algorithm, following the paper closely the code seems to be an accurate translation.

I left some small comments on the style, nothing crucial.

ShellyGarion committed:
style updates following review

gadial reviewed (dismissed):
LGTM

ShellyGarion committed:
change the default valuse of opt_a1 and opt_a2 to None

ShellyGarion committed:
add tests to the cases where opt_a1 or opt_a2 are None

ShellyGarion committed:
add release notes explaining the default behavior of opt_a1 and opt_a2

ShellyGarion committed:
minor

gadial commented on qiskit/synthesis/unitary/qsd.py:
```diff
@@ -146,6 +151,13 @@ def decomp_2q(mat):
+            opt_a2 = True
+            # check if the unitary is controlled
+            for ctrl_index in range(nqubits):
+                um00, um11, um01, um10 = _extract_multiplex_blocks(mat, ctrl_index)
```
We might call `_extract_multiplex_blocks` twice. Can this be avoided? Or is the cost negligible?

ShellyGarion replied:
I'm not sure, let's try to explore it as part of your PR :)

mtreinish replied:
It should be fairly quick, the function just moves around some axis and reshapes things. I wouldn't expect it to be a bottleneck even if we call it multiple times. The one thing we want to make sure is we're not redoing this for each recursive call though. I think it's fine since we're using the same variable, but that is what I'd be concerned about.

gadial reviewed (dismissed):
LGTM. Left one minor comment

ShellyGarion commented:
we're merging this PR, since we need it for https://github.com/Qiskit/qiskit/pull/14797.
@mtreinish @alexanderivrii - if you have any comments on the API, docs or release notes (the code itself will be ported to rust in #14797 anyway), we can handle it later.

mtreinish commented on releasenotes/notes/improve-qsd-zxz-block-decomposition-73a1a9e155ccc809.yaml:
```diff
@@ -0,0 +1,16 @@
+    (4 ** (n − 2) − 1) / 3 :class:`.CXGate`\ s.
+    With the improved decomposition, a general 3-qubit unitary can be decomposed using 19
+    :class:`.CXGate`\ s (rather than 20).
+  - |
```
This feels more like an upgrade note and not a feature. The behavior changes from 2.1 to 2.2 and we need to document it as such. I think we can get away with it because the semantics of the function are the same, the quality of the decomposition just improves by default. But you should mention that the previous default was both values were ``True`` and setting it explicitly will retain the previous behavior.

ShellyGarion replied:
changed to `upgrade_synthesis`

mtreinish commented on releasenotes/notes/improve-qsd-zxz-block-decomposition-73a1a9e155ccc809.yaml:
```diff
@@ -0,0 +1,16 @@
+    With the improved decomposition, a general 3-qubit unitary can be decomposed using 19
+    :class:`.CXGate`\ s (rather than 20).
+  - |
+    The default values of the parameters ``opt_a1`` and ``opt_a2`` of :func:`.qs_decomposition`
```
Should we deprecate these arguments?

ShellyGarion replied:
perhaps this could be done in a follow-up PR?

mtreinish replied:
Yeah we can do it in a follow up, I wasn't thinking in this PR. I was just wondering if we want to keep these arguments.

mtreinish commented on qiskit/synthesis/unitary/qsd.py:
```diff
@@ -120,52 +146,125 @@ def decomp_2q(mat):
+def _block_zxz_decomp(Umat):
+    """Block ZXZ decomposition method, by Krol and Al-Ars [2]."""
+    N = Umat.shape[0]
+    n = N // 2
```
Nit: I'd find something like this a bit easier to read. The difference between `N` and `n` was a bit confusing especially since you're explicitly saying this isn't a notation in the paper (as the comment below says the notation starts after this).

```suggestion
    dim = Umat.shape[0]
    n = dim // 2
```

mtreinish commented on qiskit/synthesis/unitary/qsd.py:
```diff
@@ -27,61 +27,79 @@
+
+    .. note::
+
+        When the original unitary is controlled then it is better to choose ``opt_a2 = False``
```
```suggestion
        When the value for the ``mat`` argument is a controlled unitary then it is better to choose ``opt_a2 = False``
```

ShellyGarion replied:
the note was removed

mtreinish commented on qiskit/synthesis/unitary/qsd.py:
```diff
@@ -27,61 +27,79 @@
+        as we do another optimization that does not work with the optimization A.2 of [1, 2].
+        When ``opt_a2 = None`` we choose the optimal value ``opt_a2 = False`` if the unitary is
+        controlled and ``opt_a2 = True`` otherwise.
+        When ``opt_a1 = None`` we choose the optimal value ``opt_a1 = True``.
```
Do we need this note anymore if you're changing the default?

ShellyGarion replied:
the note was removed

mtreinish commented on qiskit/synthesis/unitary/qsd.py:
```diff
@@ -120,52 +146,125 @@ def decomp_2q(mat):
+    if not np.allclose(Umat, Ucheck):
+        raise QiskitError(
+            "The block ZXZ decomposition matrix is not the same as the original unitary matrix."
+        )
```
Why do we need to do this? This will take an appreciable amount of time and is not user actionable is it? We can do something like this in testing (or in a rust impl as part of a debug build), but forcing it for every execution of qsd seems like a lot of overhead to check correctness for every call. I think we should move this to tests only.

ShellyGarion replied:
the verification was removed (but perhaps may be useful as part of the rust tests)

mtreinish replied:
In rust we can do a debug assertion here so when you build qiskit in debug mode we do the extra checking, but for release builds the code is stripped out.

mtreinish commented on qiskit/synthesis/unitary/qsd.py:
```diff
@@ -120,52 +146,125 @@ def decomp_2q(mat):
+
+    V, S, Wdg = scipy.linalg.svd(mat)
+    mat = V @ Wdg
+    return mat
```
Didn't we have a function for this somewhere already?

ShellyGarion replied:
we have this function in rust but not in python: https://github.com/Qiskit/qiskit/blob/2faea3849003580c69dc906c2035e743d768efa4/crates/synthesis/src/linalg/cos_sin_decomp.rs#L70
anyway, this code will be removed in Gadi's PR.

mtreinish replied:
Ok, that's probably what I'm remembering then.

ShellyGarion committed:
updates following review

mtreinish commented on qiskit/synthesis/unitary/qsd.py:
```diff
@@ -27,61 +27,70 @@
 from qiskit._accelerate.two_qubit_decompose import two_qubit_decompose_up_to_diagonal
-from qiskit._accelerate.cos_sin_decomp import cossin
+
+# pylint: disable=no-member,invalid-name, unused-variable
```
What is causing `no-member` and `unused-variable`. I think invalid-name is valid because you're using the naming from the paper? But the other two could be more concerning, are they still needed? If so maybe doing them locally would be better.

mtreinish replied:
Fixed in: https://github.com/Qiskit/qiskit/pull/14721/commits/415bb0cb16064a40a93a71748843690ac3165215

mtreinish commented on qiskit/synthesis/unitary/qsd.py:
```diff
@@ -120,52 +137,120 @@ def decomp_2q(mat):
+        mat = _closest_unitary(mat)
         circ = decomposer_2q(mat)
+        return circ
     else:
```
You can actually drop the else here and remove a layer of indentation here. I think that will make the code a bit easier to read.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/14721/commits/50194c77f856f811c294cbf9cb0b815b616688d0

mtreinish commented on qiskit/synthesis/unitary/qsd.py:
```diff
@@ -120,52 +137,120 @@ def decomp_2q(mat):
-        # perform cosine-sine decomposition
-        (u1, u2), vtheta, (v1h, v2h) = cossin(np.asarray(mat, dtype=complex))
+
+        # perform block ZXZ decomposition from [2]
```
It might have been useful to pair this to sections or formulas in the paper so it's a bit easier to follow. But since this is destined to be replaced with a rust implementation we can skip that for now

mtreinish commented on releasenotes/notes/improve-qsd-zxz-block-decomposition-73a1a9e155ccc809.yaml:
```diff
@@ -0,0 +1,16 @@
+    (4 ** (n − 2) − 1) / 3 :class:`.CXGate`\ s.
+    With the improved decomposition, a general 3-qubit unitary can be decomposed using 19
+    :class:`.CXGate`\ s (rather than 20).
+  - |
```
The first note is a feature note, the second is an upgrade note:

```suggestion
features_synthesis:
  - |
    Improve the :func:`.qs_decomposition` function which was originally based on
    Shende et. al. (https://arxiv.org/abs/quant-ph/0406176).
    The new synthesis method is based on Krol and Al-Ars (https://arxiv.org/abs/2403.13692),
    and reduce the total number of :class:`.CXGate`\ s of a general n-qubit unitary by
    (4 ** (n − 2) − 1) / 3 :class:`.CXGate`\ s.
    With the improved decomposition, a general 3-qubit unitary can be decomposed using 19
    :class:`.CXGate`\ s (rather than 20).
upgrade_synthesis:
  - |
```

mtreinish commented on releasenotes/notes/improve-qsd-zxz-block-decomposition-73a1a9e155ccc809.yaml:
```diff
@@ -0,0 +1,16 @@
+    are now ``opt_a1 = None`` and ``opt_a2 = None``.
+    When ``opt_a2 = None`` we choose the optimal value ``opt_a2 = False`` if the unitary is
+    controlled and ``opt_a2 = True`` if it is not controlled.
+    When ``opt_a1 = None`` we choose the optimal value ``opt_a1 = True``.
```
```suggestion
    The default values of the arguments ``opt_a1`` and ``opt_a2`` of :func:`.qs_decomposition`
    are now ``opt_a1 = None`` and ``opt_a2 = None``. The new defaults will choose the
    optimal value based on whether the input unitary for the ``mat`` argument is controlled
    or not to result in the minimal :class:`.CXGate` count by default. If you require the previous
    behavior you can explicitly call ``opt_a1 = True`` and ``opt_a2 = True`` to maintain the same
    behavior as previous releases.
```

mtreinish reviewed (commented):
I think this is basically ready, just a few small more comments.

mtreinish committed:
Remove unused logic to build decomposition check

In an earlier commit the internal check assertion that the decomposition
was correct was removed from the code. But the logic used to build the
check matrix was kept in place. This was not used, but also the review
comment was about the expense associated with computing the check
matrix. This commit removes the unused logic.

mtreinish committed:
Remove unnecessary pylint ignores and mark unused return variables as such

mtreinish committed:
Remove unnecessary else branch and remove indent layer

mtreinish committed:
Apply suggestions from code review

mtreinish committed:
Merge branch 'main' into qsd_zxz

mtreinish reviewed (approved):
I think this lgtm now, I pushed some small fixes to address my last comments. Thanks for the work on this
